### ISSUE TITLE
Ability to change html output

### ISSUE THREAD

**Issue by kdheepak** [tlienart/Franklin.jl#749]
If I have the current markdown file:

```markdown
+++
title = "Test"
category = "blog"
date = Date("2014-12-12")
+++

# Heading 1 without link

content

# [Heading 2 with link](https://github.com)

content
```

This is the HTML that gets generated:

```
<h1 id="heading_1_without_link">
    <a href="#heading_1_without_link">Heading 1 without link</a>
</h1>

<p>content</p>

<h1 id="heading_2_a_hrefhttpsgooglecomwith_link">
    <a href="#heading_2_a_hrefhttpsgooglecomwith_link">Heading 2 <a href="https://google.com">with link</a></a>
</h1>

<p>content</p>
```

With Tufte CSS specifically, this is how it looks:

<img width="700" alt="Screen Shot 2020-12-29 at 7 40 41 AM" src="https://user-images.githubusercontent.com/1813121/103291591-5936db00-49a9-11eb-9746-274ad9e5dfee.png">

1) It is hard to differentiate between the external link and the anchor link (this might just be resolved by changing the CSS in Tufte CSS).
2) The anchor link for the first heading is `#heading_1_without_link` but the second one is `#heading_2_a_hrefhttpsgooglecomwith_link`, which makes the links directly to sections harder to read when you want to share them with someone. If the entire heading is a link, then the anchor link is not clickable, and a user cannot guess what the anchor link to that section is going to be. I don't see a way the average user will be able to find an anchor link to this section to share with their peers.

It would be nice if I could somehow change the HTML output to the following:

```
<h1>
    <a href="#heading-1-without-link">#</a> Heading 1 without link
</h1>

<p>content</p>

<h1>
    <a href="#heading-2-with-link">#</a> <a href="https://github.com" target="_blank">Heading 2 with link</a>
</h1>

<p>content</p>
```

This looks something like this:

<img width="500" alt="Screen Shot 2020-12-29 at 7 57 18 AM" src="https://user-images.githubusercontent.com/1813121/103292660-84222e80-49ab-11eb-8825-54eccc39e2c0.png">

What I specifically wanted was the anchor link to always be visible. In this case specifically, a user can click on `#` and get the anchor link, and the content that is displayed on the webpage will always be the anchor link. And the content that is linked in the heading is differentiated from an anchor link. Also, the first heading's anchor link is `#heading-1-without-link` and the second heading's anchor link is `#heading-1-with-link`, regardless of what the link is. In my case, this is what I want to generate. But I can also imagine other users that want unique anchor links with the same title for some niche use case. It would be nice if someone the user can control what the HTML output is going to be for various different Markdown constructs.

**Comment by kdheepak** [comment]
Feel free to close this issue if you think it is out of scope for Franklin.jl.

**Comment by tlienart** [comment]
Hmm so it's interesting because a while back I was toying with the idea of exposing the logic for how headers were processed to the users. But then I thought that you could already do all of this with a lxfun or hfun and so that wasn't necessary.

Have you considered writing a function such that you could for instance write

```
\section{Foo [bar](baz) etc}
```

Which you could code as

```
function lx_section(com, _)
    content = Franklin.content(com.braces[1])
    return """~~~<h1>~~~$content~~~</h1>~~~"""
end
```

(the reason for the `~~~` is be cause the output of a `lx_*` function is re-parsed as markdown input by Franklin)

**Comment by kdheepak** [comment]
Thanks for the reply! I did look into `lxfun` or `hfun`. My understanding is that that users have to write \LaTeX like commands for that. 

I like the idea of writing \LaTeX like commands for something more specialized, but I would like to avoid it for stuff that's already very commonly used in Markdown. I know other people on the team that I'm working with (think developers + researchers writing technical documentation) will express similar opinions, and imposing the use of \LaTeX style section commands for headers would be frowned upon. 

I believe I could workaround something like this using custom Javascript though. I guess I just wanted to check if this would be something you are interested in. 

I have a few other things like this, where I'd like to input to be standard or close to standard markdown but the output to be HTML that is different from what Franklin currently generates. Here's one more example:

```
This has a footnote[^1]

[^1]: footnote definition
```

This is the syntax Franklin currently supports for footnotes. I like using this syntax in Markdown documents. 

However in Tufte CSS, there's two kinds of "footnotes", footnotes with numbers and margin notes without numbers. See <https://edwardtufte.github.io/tufte-css/#sidenotes> for more information.

For my blog, I use these interchangeably in a single post a lot when writing. See example of both in this post: https://blog.kdheepak.com/my-unicode-cheat-sheet.html#vim 

Here's an example of a margin note in the source.

```
This emoji actually consists of 5 code points[^uniview]:

[^uniview]: _aside_: We can view this breakdown using uniview.
```

The way I had set it up is that if I have a "footnote" that starts with the text ` _aside_: `, it gets modifies by a Pandoc filter I wrote and generates the HTML for Tufte CSS margin notes instead. Any footnote without the ` _aside_: ` is left untouched and that ends up being the correct the HTML for numbered footnotes.

I can reversibly change between using margin notes and numbered side notes, just by adding or removing the ` _aside_: `. 

If I did this in Franklin, I wouldn't want to use `\marginnote{We can view this breakdown using uniview.}` versus `[^We can view this breakdown using uniview.]` for the margin notes versus footnotes. 

This is probably a more complicated and niche request and maybe pertains just to my blog / Tufte CSS, but I think it generally falls under the category of parsing default Markdown to generate custom HTML.

**Comment by tlienart** [comment]
Right, I understand.

I think for the headers I'm potentially inclined to expose this and let the users specify their approach somehow, for the footnote/margin note what you want will likely not be supported soon as this would also require additions to the parser. I'm currently working on the parser and there may be a possibility for users to write their own additions to it but not soon.

Note that I understand that you don't like the latex or hfun syntax but this is the de facto way Franklin users can change the HTML that gets produced at will... this will not change and I'm not inclined to add case by case specific features which can already be coded that way since it will make maintenance much harder.

**Comment by kdheepak** [comment]
Great! Again, feel free to close this if it is not in the scope of Franklin. I do think this is a rather niche request, and most users would be happy with `hfun` / `lxfun` functionality.

Also, it's not that I don't like LaTeX like syntax :) I think that functionality is pretty neat and I would in fact prefer it to an arbitrary alternative syntax. It's just that for well established Markdown syntax that I (and others) often use, I have a preference to avoid that LaTeX style and stick to Markdown.

**Comment by kdheepak** [comment]
Alternatively, if it is possible to add functionality for creating a plugin for extending the parser or using an alternative parser, I'd be willing to write and maintain a Franklin plugin to handle that. Maybe you have ideas for this but when I went through the source code I didn't find a straight forward way to hook into the parser at this point.

**Comment by tlienart** [comment]
Yes I think that will be made available, no promises as to when as I'm doing pretty significant refactoring of the parser to use CommonMark.jl as backend and allow pdf output but I think allowing users to specify how to extend the parser could be useful to some.

Thanks again for the feedback! I'll close this for now but please don't hesitate to open new issues with comments and suggestions, not sure I can always address them straightaway but IM keen to make things easy for users to modify themselves 😁

**Comment by tlienart** [comment]
It occurred to me that for the header thing, all you actually need is to be able to toggle off the fact that Franklin makes headers into links; this would be pretty easy to do, and then after that you'd just put your own links in there.

The footnote one is non standard markdown afaik so that would go with the parser extensions that we discussed earlier.
